### PR TITLE
[KYUUBI #4788][K8S][HELM] Use StatefulSet instead of Deployment

### PR DESCRIPTION
**Pull request by camper42**
### _Why are the changes needed?_

Use StatefulSet instead of Deployment, add a headless service for statefulset

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

image
image

### CODE REVIEW

File: charts/kyuubi/templates/kyuubi-statefulset.yaml [hunk 1]
```diff
@@ -71,7 +77,7 @@ spec:
           {{- if .Values.livenessProbe.enabled }}
           livenessProbe:
             exec:
-              command: ["/bin/bash", "-c", "bin/kyuubi status"]
+              command: [ "/bin/bash", "-c", "bin/kyuubi status" ]
```
Comment (pan3793) [inline charts/kyuubi/templates/kyuubi-statefulset.yaml:80]:
```suggestion
              command: ["/bin/bash", "-c", "bin/kyuubi status"]
```

**Review by pan3793 — APPROVED** [review]
LGTM, @dnskr would you mind taking a look?

**Comment by camper42** [comment]
btw, our production kyuubi changed to this version

**Comment by pan3793** [comment]
@camper42 thanks for sharing, cc @turboFei @zwangsheng too

File: charts/kyuubi/templates/kyuubi-service.yaml [hunk 2]
```diff
@@ -15,6 +15,23 @@
...
 */}}
 
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ $.Release.Name }}-headless
+  labels:
+    {{- include "kyuubi.labels" $ | nindent 4 }}
+spec:
+  type: ClusterIP
```
Comment (zwangsheng) [inline charts/kyuubi/templates/kyuubi-service.yaml:25]: we should add `clusterIp: None` here, as kubernetes doc said:

>  In this case, you can create what are termed headless Services, by explicitly specifying "None" for the cluster IP address (.spec.clusterIP).

Comment (camper42) [reply charts/kyuubi/templates/kyuubi-service.yaml:25]: thanks to point that, I have to delete the headless svc to fix this.
luckily, there is no direct use of this headless service at the moment.

**Review by zwangsheng — APPROVED** [review]
LGTM

**Review by dnskr — CHANGES_REQUESTED** [review]
I left few requests for changes.

I'm not sure I know why these changes are needed, so it would be useful for me if you could provide more details about the PR:
 - Why is it needed to replace `Deployment` by `StetefullSet`, i.e. what kind of problem does it solve?
 - Would it be enough to use headless Service and Deployment?
 - Which Service (headless or not) clients should connect to and use?

**Comment by camper42** [comment]
> - Why is it needed to replace Deployment by StetefullSet, i.e. what kind of problem does it solve?
> - Would it be enough to use headless Service and Deployment?

as our deployment, we want to persist server log later, so statefulset + pvc maybe a better choice

and ref to #4788: Kyuubi server is actually stateful for batch API, especially the batch session recovery relies on stable IP address/hostname after restarting Kyuubi server

> Which Service (headless or not) clients should connect to and use?

at the moment headless service only used to provide stable network ID. Our client is not in the same cluster as the kyuubi server, in this case we use kyuubi-thrift-binary's(LoadBalancer) external IP.

And I'm not sure which service it better for in-cluster client connect. maybe @pan3793 know more about it.

- client -> DNS: kyuubi-headless.kyuubi.svc.cluster.local -> pod IP (may have DNS cache issue)
- client -> kyuubi-protocol cluster IP -> pod ?

---


https://kyuubi.readthedocs.io/en/v1.7.1-rc0/deployment/high_availability_guide.html#client-side-usage
> then it can randomly pick one of the Kyuubi service uris from the specified ZooKeeper addresses in the /kyuubi path.

in ZK HA mode, client direct connect to a randomly picked kyuubi server.

In kubernetes, whether we use headless services (DNS service discovery) or ClusterIP (load balancing), it does not involve zk, so can HA be achieved without zk?

**Comment by pan3793** [comment]
> I'm not sure which service it better for in-cluster client connect.

For THTIFT_BINARY (TCP) it does not matter, because it's a long connection, the whole lifecycle of a session while sticky in one Kyuubi Server Pod.

For REST (HTTP), I think headless should be better. In the batch submission (aka Spark Jar application), suppose the `spark-submit` was executed in the Kyuubi-Server-1, the following request to fetch local logs of `spark-submit` process must be handled by the same Pod, the current implementation is:

1. Client => [Kyuubi Server Service] => random Kyuubi Server Pod
2.1 If the Kyuubi Server Pod holds the batch session => handle and return response
2.2 If the Kyuubi Server Pod does not hold the batch session, check the state store (and external storage like MySQL) to find which Kyuubi Server holds it and forward the request to it, wait for the response, then pass through it to the Client

The 2.2 could be enhanced by
- the response contains the Kyuubi-Server-1 Pod address
- the client always(until network exception) tries to access the Kyuubi-Server-1 directly to avoid Kyuubi Server internal forwarding overhead.

In such a case, I would say Headless Service seems more suitable.

**Comment by camper42** [comment]
headless service `kyuubi-headless.kyuubi.svc.cluster.local` will resolve to:
- Pod 0 IP
- Pod 1 IP
- ...
- Pod N IP

---

Pod DNS(stable network ID) will resolve to Pod IP:
- kyuubi-0.kyuubi-headless.kyuubi.svc.cluster.local -> Pod 0 IP
- kyuubi-1.kyuubi-headless.kyuubi.svc.cluster.local -> Pod 1 IP
- ...
- kyuubi-N.kyuubi-headless.kyuubi.svc.cluster.local -> Pod N IP

---

Cluster IP/External IP will proxy to one Pod IP (and stick for a client IP if sessionAffinity enabled)

---

in the current implementation, no difference between service expose methods for kyuubi REST ?

> - the response contains the Kyuubi-Server-1 Pod address
> - the client always(until network exception) tries to access the Kyuubi-Server-1 directly to avoid Kyuubi Server internal forwarding overhead.

in this implementation, client needs to reconnect to the Kyuubi-Server-1 Pod address, if using HTTP, client can connect by Pod IP address, but if using HTTPS, we need a stable address?

**Comment by pan3793** [comment]
@turboFei could you share some ideas about HTTPS and the Service type chosen?

**Review by dnskr — APPROVED** [review]
LGTM
@camper42 thanks for your contribution and sorry for the delay with review.

**Comment by turboFei** [comment]
> @turboFei could you share some ideas about HTTPS and the Service type chosen?


we are using stateful set.

A stable FQDN is needed for kyuubi batch recovery.

For HTTPS, stateful set is not a must.

**Comment by pan3793** [comment]
Thanks, merged to master

**Comment by pan3793** [comment]
Supply a new advantage of `StatefulSet` over `Deployment`, `Utils.findLocalInetAddress.getCanonicalHostName` returns different values:

- `Deployment`: Pod name, e.g. `kyuubi-bd8f56ddb-59jn5`
- `StatefulSet`: FQDN, e.g. `kyuubi-server-0.kyuubi-headless.spark.svc.cluster.local`

Thus, switching to `StatefulSet` could overcome the errors like 

```
Caused by: java.net.UnknownHostException: kyuubi-bd8f56ddb-59jn5
```

when `kyuubi.frontend.connection.url.use.hostname=true` (default is `true`)

**Pull request closed without merging**